# IPython magic function documentation - `%watermark`

I wrote this simple `watermark` IPython magic function to conveniently add date- and time-stamps to my IPython notebooks. Also, I often want to document various system information, e.g., for my [Python benchmarks](https://github.com/rasbt/One-Python-benchmark-per-day) series.


## Installation

The watermark line magic can be installed by executing

```bash
pip install watermark
```

Alternatively, you can install the latest development version directly from GitHub via

```bash
pip install -e git+https://github.com/rasbt/watermark#egg=watermark
```

## Loading the `%watermark` magic

To load the `watermark` magic, execute the following line in your IPython notebook or current IPython shell

In [1]:
%load_ext watermark

The version of `watermark` itself can be retrieved as follows:

In [2]:
%watermark --watermark

Watermark: 2.1.0



## Usage

In order to display the optional `watermark` arguments, type

In [3]:
%watermark?

Docstring:
::

  %watermark [-a AUTHOR] [-d] [-n] [-t] [-i] [-z] [-u] [-c CUSTOM_TIME]
                 [-v] [-p PACKAGES] [-h] [-m] [-g] [-r] [-b] [-w] [-iv]

IPython magic function to print date/time stamps
and various system information.

optional arguments:
  -a AUTHOR, --author AUTHOR
                        prints author name
  -d, --date            prints current date as YYYY-mm-dd
  -n, --datename        prints date with abbrv. day and month names
  -t, --time            prints current time as HH-MM-SS
  -i, --iso8601         prints the combined date and time including the time
                        zone in the ISO 8601 standard with UTC offset
  -z, --timezone        appends the local time zone
  -u, --updated         appends a string "Last updated: "
  -c CUSTOM_TIME, --custom_time CUSTOM_TIME
                        prints a valid strftime() string
  -v, --python          prints Python and IPython version
  -p PACKAGES, --packages PACKAGES
                        prints ve

<pre>
%watermark [-a AUTHOR] [-d] [-n] [-t] [-i] [-z] [-u] [-c CUSTOM_TIME]
                 [-v] [-p PACKAGES] [-h] [-m] [-g] [-w]

IPython magic function to print date/time stamps
and various system information.

optional arguments:
  -a AUTHOR, --author AUTHOR
                        prints author name
  -d, --date            prints current date as YYYY-mm-dd
  -n, --datename        prints date with abbrv. day and month names
  -t, --time            prints current time as HH-MM-SS
  -i, --iso8601         prints the combined date and time including the time
                        zone the ISO 8601 standard with UTC offset
  -z, --timezone        appends the local time zone
  -u, --updated         appends a string "Last updated: "
  -c CUSTOM_TIME, --custom_time CUSTOM_TIME
                        prints a valid strftime() string
  -v, --python          prints Python and IPython version
  -p PACKAGES, --packages PACKAGES
                        prints versions of specified Python modules and
                        packages
  -h, --hostname        prints the host name
  -m, --machine         prints system and machine info
  -g, --githash         prints current Git commit hash
  -w, --watermark       prints the current version of watermark
</pre>

## Examples

### Defaults

In [4]:
%watermark

Last updated: 2020-11-20T15:27:00.378508+00:00

Python implementation: CPython
Python version       : 3.7.8
IPython version      : 7.16.1

Compiler    : GCC 7.5.0
OS          : Linux
Release     : 4.19.112+
Machine     : x86_64
Processor   : x86_64
CPU cores   : 8
Architecture: 64bit



### Last updated date and time

In [5]:
%watermark -u -t -d

Last updated: 2020-11-20 15:27:00



In [6]:
%watermark -u --iso8601

Last updated: 2020-11-20T15:27:00.417982+00:00



In [7]:
%watermark -u -n -t -z

Last updated: Fri Nov 20 2020 15:27:00UTC



### Python version information

In [8]:
%watermark -v

Python implementation: CPython
Python version       : 3.7.8
IPython version      : 7.16.1



### Machine information

In [9]:
%watermark -m

Compiler    : GCC 7.5.0
OS          : Linux
Release     : 4.19.112+
Machine     : x86_64
Processor   : x86_64
CPU cores   : 8
Architecture: 64bit



### Combination of information

In [10]:
%watermark -v -m -p numpy,scipy,sklearn -g

Python implementation: CPython
Python version       : 3.7.8
IPython version      : 7.16.1

numpy  : 1.19.4
scipy  : 1.5.4
sklearn: 0.23.2

Compiler    : GCC 7.5.0
OS          : Linux
Release     : 4.19.112+
Machine     : x86_64
Processor   : x86_64
CPU cores   : 8
Architecture: 64bit

Git hash: b9768dfeebb4a079fc1a7b8d83f249ded3f31ad8



### With author

In [11]:
%watermark -a "John Doe" -u -d -v -m

Author: John Doe

Last updated: 2020-11-20

Python implementation: CPython
Python version       : 3.7.8
IPython version      : 7.16.1

Compiler    : GCC 7.5.0
OS          : Linux
Release     : 4.19.112+
Machine     : x86_64
Processor   : x86_64
CPU cores   : 8
Architecture: 64bit



### Imported modules information

In [12]:
import scipy as sp
from sklearn import metrics
import numpy.linalg as linalg

In [13]:
%watermark --iversions

numpy  : 1.19.4
sklearn: 0.23.2
scipy  : 1.5.4

